nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -737,3 +740,715 @@ async fn unlock_instance_inner(
+        let nexus = &cptestctx.server.server_context().nexus;
+        nexus
+            .sagas
+            .saga_execute::<SagaInstanceUpdate>(test.saga_params())
+            .await
+            .expect("update saga should succeed");
+
+        test.verify_src_succeeded(cptestctx).await;
+    }
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_migration_source_completed_actions_succeed_idempotently(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        let _project_id = setup_test_project(&cptestctx.external_client).await;
+        let other_sleds = test_helpers::add_sleds(cptestctx, 1).await;
+        let test = MigrationTest::setup(cptestctx, &other_sleds).await;
+
+        // Pretend the migration source has completed.
+        test.update_src_state(
+            cptestctx,
+            VmmState::Stopping,
+            MigrationState::Completed,
+        )
+        .await;
+
+        // Build the saga DAG with the provided test parameters
+        let dag =
+            create_saga_dag::<SagaInstanceUpdate>(test.saga_params()).unwrap();
+
+        crate::app::sagas::test_helpers::actions_succeed_idempotently(
+            &cptestctx.server.server_context().nexus,
+            dag,
+        )
+        .await;
+
+        test.verify_src_succeeded(cptestctx).await;
+    }
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_migration_source_completed_can_unwind(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        let nexus = &cptestctx.server.server_context().nexus;
+        let other_sleds = test_helpers::add_sleds(cptestctx, 1).await;
+        let _project_id = setup_test_project(&cptestctx.external_client).await;
+
+        test_helpers::action_failure_can_unwind::<SagaInstanceUpdate, _, _>(
+            nexus,
+            || {
+                Box::pin(async {
+                    let test =
+                        MigrationTest::setup(cptestctx, &other_sleds).await;
+                    // Pretend the migration source has completed.
+                    test.update_src_state(
+                        cptestctx,
+                        VmmState::Stopping,
+                        MigrationState::Completed,
+                    )
+                    .await;
+                    test.saga_params()
+                })
+            },
+            || {
+                Box::pin({
+                    async {
+                        let state = test_helpers::instance_fetch_by_name(
+                            cptestctx,
+                            INSTANCE_NAME,
+                            PROJECT_NAME,
+                        )
+                        .await;
+                        let instance = state.instance();
+
+                        // Unlike most other sagas, we actually don't unwind the
+                        // work performed by an update saga, as we would prefer
+                        // that at least some of it succeeds. The only thing
```

Comment:
In general the tests in this module look great to me--thanks for going the extra mile to cover so many different cases!

In addition to asserting that the instance is unlocked here, I think it's worth mulling over what post-states we expect *not* to have when one of these sagas unwinds. The main one I can think of offhand is along the lines of "instance is in NoVmm but has a valid destination Propolis ID," but with some thought I suspect I could come up with others. (The general idea would be to look not for cases where the saga didn't undo a step correctly, but for cases where some update should have been atomic but was not.)